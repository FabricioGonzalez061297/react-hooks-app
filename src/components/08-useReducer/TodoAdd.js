import React from 'react'

import {UseForm } from './../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [{ description }, handleInputChange, reset] = UseForm({
        description: ''
    });

    const handleSubmit =(e)=>{
        e.preventDefault();

        if (description.trim().length <= 0) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc:description,
            done:false
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>

            <h5>Agregar TODO</h5>
            <hr></hr>

            <form onSubmit={handleSubmit} >

                <input
                    type="text"
                    name="description"
                    placeholder="Aprender ..."
                    autoComplete="off"
                    className="form-control"
                    onChange={handleInputChange}
                    value={description}

                />

                <div className="d-grid gap-2">
                    <button
                        type="submit"
                        className="btn btn-outline-primary mt-1 "
                    >
                        Agregar
                        </button>
                </div>


            </form>


        </>
    )
}
