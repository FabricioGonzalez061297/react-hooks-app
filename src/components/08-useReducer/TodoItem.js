import React from 'react'

export const TodoItem = ({todo,handleToggle,handleDelete,i}) => {
    return (
        <li
            className="list-group-item"
             >
            <p onClick={() => handleToggle(todo.id)} className={`${todo.done && 'complete'}`} > {i + 1}.  {todo.desc}</p>

            <button
                onClick={() => handleDelete(todo.id)}
                className="btn btn-danger mt-2"
            >Borrar</button>
        </li>
    )
}
