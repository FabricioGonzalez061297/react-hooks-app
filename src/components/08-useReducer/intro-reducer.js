console.log('hola mundo');


const initialState = [
    {
        id: 1,
        todo: "comprar leche",
        done: false
    }
];


const todoReducer = (state = initialState, action) => {

    switch (action?.type) {
        case 'agregar':{
            return [...state,action.payload];
        }
        default:{
            return state;
        }
        
    }

}

let todos= todoReducer();


const newTodo = {
    id: 2,
    todo: "comprar Agua",
    done: false
};

const addTodo= {
    type:'agregar',
    payload:newTodo
}

todos=todoReducer(todos,addTodo);


console.log(todos)
