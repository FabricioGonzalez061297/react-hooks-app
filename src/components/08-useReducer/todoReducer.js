export const todoReducer=(state=[],action)=>{

    switch (action?.type) {
        case 'add':{
            return [...state,action.payload];
        }

        case 'delete':{

            const newState= state.filter(t=> t.id !== action.payload);

            return [...newState];
        }

        case 'toggle':{
            return state.map(t=> t.id===action.payload 
                ? {...t, done:!t.done}
                : t );
        }

        case 'toggle-old':{
            return state.map( t=>{

                if(t.id===action.payload){
                    t.done=!t.done;
                }
                return t;
            });
        }
    
        default:{
            return state;
        }
    }

}