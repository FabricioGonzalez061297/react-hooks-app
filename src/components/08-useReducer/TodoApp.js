import React, { useReducer, useEffect } from 'react';
import { todoReducer } from './todoReducer';

import './styles.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';


const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

   
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);



    const handleDelete = (todoId) => {
        if (!todoId) return;

        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch(action);
    }

    const handleToggle = (todoId) => {
        dispatch({ type: 'toggle', payload: todoId });
    }

    const handleAddTodo= (newTodo)=>{
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

 

    return (
        <div>
            <h1>TodoApp {todos.length} </h1>
            <hr></hr>


            <div className="row" >

                <div
                    className="col-md-7 col-sm-12"
                >
                    <TodoList handleDelete={handleDelete} handleToggle={handleToggle} todos={todos} />

                </div>


                <div className="col-md-5 col-sm-12" >

                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>

            </div>


        </div>
    )
}
