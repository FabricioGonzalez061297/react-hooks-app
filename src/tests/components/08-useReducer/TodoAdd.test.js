import { shallow } from "enzyme"
import {TodoAdd} from './../../../components/08-useReducer/TodoAdd';

const handleAddTodo= jest.fn();

describe('Pruebas en <TodoAdd/>',()=>{

    test('debe de mostrarse correctamente', () => {
      
        const wrapper= shallow(
        <TodoAdd
            handleAddTodo={handleAddTodo}
        />);

        expect(wrapper).toMatchSnapshot();
        
    });

    test('No debe de llamar addTODO', () => {
      
        const wrapper= shallow(
        <TodoAdd
            handleAddTodo={handleAddTodo}
        />);

        const formSubmit= wrapper.find('form').prop('onSubmit');

        formSubmit({preventDefault(){}});
        
        expect(handleAddTodo).toHaveBeenCalledTimes(0);

    });

    test('debe de llamar la funcion handleAddTodo', () => {
      
        const value= 'Aprender react';

        const wrapper= shallow(
        <TodoAdd
            handleAddTodo={handleAddTodo}
        />);

        wrapper.find('input').simulate('change',{
            target:{
                value,
                name:'description'
            }
        });

        const formSubmit= wrapper.find('form').prop('onSubmit');

        formSubmit({preventDefault(){}});
        
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith({
            id:expect.any(Number),
            desc:value,
            done:false
        });

        expect(wrapper.find('input').prop('value')).toBe('');

    });
    
})