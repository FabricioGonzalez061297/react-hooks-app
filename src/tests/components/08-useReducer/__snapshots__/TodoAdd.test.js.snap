// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd/> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h5>
    Agregar TODO
  </h5>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender ..."
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-outline-primary mt-1 "
        type="submit"
      >
        Agregar
      </button>
    </div>
  </form>
</Fragment>
`;
